### PR TITLE
StateMachine: Add test/example for per-transfer balance invariant

### PR DESCRIPTION
**Pull request by sentientwaffle**
Add a test which applies a balance invariant for a single transfer -- that is, temporarily apply a balance invariant to an account that doesn't have a balance invariant.

### CODE REVIEW

File: src/state_machine.zig [hunk 1]
```diff
@@ -4520,6 +4520,33 @@ test "create_transfers: multiple debits, single credit, balancing debits" {
...
+        \\ transfer T3 A2 A3  0  T2  _  _  _    0 L1 C1   _   _ _ VOI   _   _ _ _ _ _ _ ok
+        \\ commit create_transfers
+        \\
+        \\ lookup_account A1 0 40 0  0 _
+        \\ lookup_account A2 0 40 0 40 _
+        \\ lookup_account A3 0  0 0  0 _
+        \\ commit lookup_accounts
+    );
+}
+
```
Comment (batiati) [inline src/state_machine.zig]: Can we add the reverse example in this same test?

```zig
    // Temporarily enforce `debits_must_not_exceed_credits` on account `A1`.
    try check(
        \\ account A1  0  0  0  0  _  _  _ _ L1 C1   _   _   _ _ _ _ _ _ ok
        \\ account A2  0  0  0  0  _  _  _ _ L1 C1   _   _   _ _ _ _ _ _ ok
        \\ account A3  0  0  0  0  _  _  _ _ L1 C1   _   D<C _ _ _ _ _ _ ok
        \\ commit create_accounts
        \\
        \\ setup A1 0 0 0 40
        \\
        \\ transfer T1 A1 A2 41   _  _  _  _    0 L1 C1 LNK   _ _   _   _   _ _ _ _ _ _ linked_event_failed
        \\ transfer T2 A3 A1 41   _  _  _  _    0 L1 C1 LNK PEN _   _   _ BCR _ _ _ _ _ exceeds_credits
        \\ transfer T3 A3 A1  0  T2  _  _  _    0 L1 C1   _   _ _ VOI   _   _ _ _ _ _ _ linked_event_failed
        \\ commit create_transfers
        \\
        \\ transfer T1 A1 A2 40   _  _  _  _    0 L1 C1 LNK   _ _   _   _   _ _ _ _ _ _ ok
        \\ transfer T2 A3 A1 40   _  _  _  _    0 L1 C1 LNK PEN _   _   _ BCR _ _ _ _ _ ok
        \\ transfer T3 A3 A1  0  T2  _  _  _    0 L1 C1   _   _ _ VOI   _   _ _ _ _ _ _ ok
        \\ commit create_transfers
        \\
        \\ lookup_account A1 0 40 0 40 _
        \\ lookup_account A2 0  0 0 40 _
        \\ lookup_account A3 0  0 0  0 _
        \\ commit lookup_accounts
    );
```

**Review by batiati — COMMENTED** [review]
It would be great to add it to the recipes!

**Comment by sentientwaffle** [comment]
Also simplified this a little bit -- per https://github.com/tigerbeetle/tigerbeetle/pull/2327/commits/f5e51ee6d6136ce96d2dd649c7437248f7a5384f, the pending transfer can just set `amount=1`.

File: docs/coding/recipes/balance-invariant-transfers.md [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+  [`flags.debits_must_not_exceed_credits`](../../reference/account.md#flagsdebits_must_not_exceed_credits) set,
+  since in this recipe we are only enforcing the invariant on a per-transfer basis.
+3. The **control** account, to test the balance invariant. The control account should have
+  [`flags.credits_must_not_exceed_debits`](../../reference/account.md#flagscredits_must_not_exceed_debits)
+  set.
+
+| Id | Debit Account | Credit Account | Amount | Pending Id |                                               Flags |
+| -: | ------------: | -------------: | -----: | ---------: | --------------------------------------------------: |
+|  1 |        Source |    Destination |    123 |          - | [`linked`](../../reference/transfer.md#flagslinked) |
+|  2 |   Destination |        Control |      1 |          - | [`linked`](../../reference/transfer.md#flagslinked), [`pending`](../../reference/transfer.md#flagspending) |
```
Comment (batiati) [inline docs/coding/recipes/balance-invariant-transfers.md]: We should mention the flags `balancing_credit` and `balancing_debit` here.

Comment (sentientwaffle) [reply docs/coding/recipes/balance-invariant-transfers.md]: Ah, good catch!

**Review by batiati — APPROVED** [review]
great chess move! ♞

**Pull request merged**